### PR TITLE
C#: Add more `environment` and `commandargs` sources for the C# Standard Library

### PR DESCRIPTION
**Pull request by egregius313**
Adds more models for the `environment` and `commandargs` local source kinds.

This primarily focuses on the .NET standard library and the `Microsoft.Extensions.Configuration` library.

### CODE REVIEW

File: csharp/ql/test/resources/stubs/Microsoft.Extensions.Configuration/8.0.0/Microsoft.Extensions.Configuration.cs [hunk 1]
```diff
@@ -0,0 +1,102 @@
+namespace Microsoft.Extensions.Configuration
```
Comment (michaelnebel) [inline csharp/ql/test/resources/stubs/Microsoft.Extensions.Configuration/8.0.0/Microsoft.Extensions.Configuration.cs:1]: The subfolders of `stubs` primary contains auto generated stubs and the files in the stubs folder are "handwritten".
In `csharp/scripts/stubs` there are scripts for generating stubs from NuGet packages.
However, in this case I think there already exists stubs for these methods. They are in the ASP.NET framework instead.

File: csharp/ql/test/library-tests/dataflow/flowsources/local/environment/options [hunk 2]
```diff
@@ -0,0 +1,4 @@
+semmle-extractor-options: /nostdlib /noconfig
+semmle-extractor-options: --load-sources-from-project:${testdir}/../../../../../resources/stubs/_frameworks/Microsoft.NETCore.App/Microsoft.NETCore.App.csproj
```
Comment (michaelnebel) [inline csharp/ql/test/library-tests/dataflow/flowsources/local/environment/options:2]: Should point to the ASP.NET framework instead.

File: csharp/ql/test/library-tests/dataflow/flowsources/local/environment/options [hunk 3]
```diff
@@ -0,0 +1,4 @@
...
+semmle-extractor-options: --load-sources-from-project:${testdir}/../../../../../resources/stubs/Microsoft.Extensions.Configuration/8.0.0/Microsoft.Extensions.Configuration.csproj
```
Comment (michaelnebel) [inline csharp/ql/test/library-tests/dataflow/flowsources/local/environment/options:3]: Can be deleted.

**Review by michaelnebel — COMMENTED** [review]
Thank you!
I have added some comments.
Could you elaborate a bit on, how you found the sources?

File: csharp/ql/lib/ext/Microsoft.Extensions.Configuration.model.yml [hunk 4]
```diff
@@ -0,0 +1,15 @@
+extensions:
+  - addsTo:
+      pack: codeql/csharp-all
+      extensible: sourceModel
+    data:
+      - ["Microsoft.Extensions.Configuration", "EnvironmentVariablesExtensions", False, "AddEnvironmentVariables", "(Microsoft.Extensions.Configuration.IConfigurationBuilder)", "", "ReturnValue", "environment", "manual"]
```
[the remark this answers is not in the document]
Comment (egregius313) [reply csharp/ql/lib/ext/Microsoft.Extensions.Configuration.model.yml:6]: > Also, out of curiosity - how did you come across these?

I was looking into packages under the `System.*` and `Microsoft.*` namespaces (.NET standard library and Microsoft libraries) for APIs dealing with environment variables and/or command line arguments. And in the process, I found `Microsoft.Extensions.Configuration`.

**Comment by egregius313** [comment]
> Could you elaborate a bit on, how you found the sources?

I was looking into packages under the `System.*` and `Microsoft.*` namespaces (.NET standard library and Microsoft libraries) for APIs dealing with environment variables and/or command line arguments. And in the process, I found `Microsoft.Extensions.Configuration`.

The standard library models were mostly from looking into the standard way to access command line arguments and environment variables with `System.Environment`.

**Comment by egregius313** [comment]
@michaelnebel I have modified the `Microsoft.Extensions.Configuration` models to handle the fluent API, and have added taint flow tests for the extension methods to check that things are being tracked appropriately with both versions of the fluent API.

File: csharp/ql/lib/ext/Microsoft.Extensions.Configuration.model.yml [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    data:
+      - ["Microsoft.Extensions.Configuration", "EnvironmentVariablesExtensions", False, "AddEnvironmentVariables", "", "", "Argument[0]", "environment", "manual"]
+      - ["Microsoft.Extensions.Configuration", "EnvironmentVariablesExtensions", False, "AddEnvironmentVariables", "", "", "ReturnValue", "environment", "manual"]
+  - addsTo:
+      pack: codeql/csharp-all
+      extensible: summaryModel
+    data:
+      - ["Microsoft.Extensions.Configuration", "IConfiguration", True, "get_Item", "(System.String)", "", "Argument[this]", "ReturnValue", "taint", "manual"]
+      - ["Microsoft.Extensions.Configuration", "IConfigurationBuilder", True, "Build", "()", "", "Argument[this]", "ReturnValue", "taint", "manual"]
+      - ["Microsoft.Extensions.Configuration", "CommandLineConfigurationExtensions", False, "AddCommandLine", "", "", "Argument[1..2]", "Argument[0]", "taint", "manual"]
```
Comment (michaelnebel) [inline csharp/ql/lib/ext/Microsoft.Extensions.Configuration.model.yml:14]: Hmm, I just took another look at the type signature for the `AddCommandLine` methods (at least those we have stubs for) and we can't do a forall quantification on `Argument[2]` (as they don't all have three parameters). So we need to target the method with three parameters (this should also remove the failing consistency checks).

File: csharp/ql/test/library-tests/dataflow/flowsources/local/commandargs/CommandArgs.ql [hunk 6]
```diff
@@ -0,0 +1,6 @@
+import csharp
```
Comment (michaelnebel) [inline csharp/ql/test/library-tests/dataflow/flowsources/local/commandargs/CommandArgs.ql:1]: The expected output file needs to be added.

File: csharp/ql/test/library-tests/dataflow/flowsources/local/environment/EnvironmentVariableFlow.ql [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import csharp
+import semmle.code.csharp.dataflow.internal.ExternalFlow
+
+module EnvironmentVariableFlowConfig implements DataFlow::ConfigSig {
+  predicate isSource(DataFlow::Node source) { sourceNode(source, "environment") }
```
Comment (michaelnebel) [inline csharp/ql/test/library-tests/dataflow/flowsources/local/environment/EnvironmentVariableFlow.ql:5]: nit: You can consider to do this using threat models (also fine to leave it as it is).

**Comment by michaelnebel** [comment]
> Excellent! I will start a DCA run; Also we need to hold the merge of this PR as one of the queries now relies on threat models.

Alternatively split the PR in two such that it is follow up PR that changes the semantics of the query.

**Comment by michaelnebel** [comment]
> > Excellent! I will start a DCA run; Also we need to hold the merge of this PR as one of the queries now relies on threat models.
> 
> Alternatively split the PR in two such that it is follow up PR that changes the semantics of the query.

Looks like there is no need for this.
I think that we can start to merge all the threat model PRs.

**Pull request merged**